### ISSUE TITLE
ConcurrentWrites  with .Net 5 and .Net 4.8 Modules

### ISSUE THREAD

**Issue by heebtob** [NLog/NLog#4242]
Is it possible to have multiple processes (.Net 5 and .Net 4.8) writing to the same logfile (FileTarget) with a global mutex for performance reasons?
So far we had only .Net 4.8 Modules but now we begin with migrating Modules to .Net 5.

**Comment by snakefoot** [comment]
.Net48 on Windows will perform atomic-filewrites without global-mutex by default (much faster).

.Net50 with NetStandard will use global-mutex by default (because atomic filewrites are not available).

They should use the same globa-mutex-name if you configure FileTarget to use the following options on both .Net48 + .Net50 platform:
- KeepFileOpen = true
- ConcurrentWrites = true
- ForceMutexConcurrentWrites = true

**Comment by heebtob** [comment]
Many thanks for answering so fast. I will try it out with these options.

**Comment by heebtob** [comment]
Unfortunately this does not always work but I narrowed down the problem.
I have .Net 4.8 Modules with IIS and Windows Services with .Net 5.

If IIS starts first the .Net 4.8 Modules create the global Mutex with Everyone Rights.
Then I start the .Net 5 Services and they start logging and everything works fine.

If I start the .Net 5 Services first it creates a global Mutex with System Rights and after that the .Net 4.8 Modules on IIS the .Net 4.8 Modules do not log.
It gives me Exceptions in the internal NLog Trace.

2021-01-18 18:40:11.2805 Warn Failed to create global archive mutex: C:\RCWork\Traces\RecosIC\\RL-DEV61.log Exception: System.UnauthorizedAccessException: Access to the path 'Global\NLog-FileFileArchiveLock-c:_rcwork_traces_recosic_rl-dev61.log' is denied.
 at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
 at System.Threading.Mutex.MutexTryCodeHelper.MutexTryCode(Object userData)
 at System.Runtime.CompilerServices.RuntimeHelpers.ExecuteCodeWithGuaranteedCleanup(TryCode code, CleanupCode backoutCode, Object userData)
 at System.Threading.Mutex.CreateMutexWithGuaranteedCleanup(Boolean initiallyOwned, String name, Boolean& createdNew, SECURITY_ATTRIBUTES secAttrs)
 at System.Threading.Mutex..ctor(Boolean initiallyOwned, String name, Boolean& createdNew, MutexSecurity mutexSecurity)
 at NLog.Internal.FileAppenders.BaseMutexFileAppender.ForceCreateSharableMutex(String name)
 at NLog.Internal.FileAppenders.BaseMutexFileAppender.CreateArchiveMutex()

Then I checked the mutex security with WinObj and indeed there is a difference.
.Net 4.8 creates the Mutex with the "Everyone" Access Rights and .Net Core creates the Mutex with inherited Access Rights.
I know that .Net Standard does not support the Constructor Mutex(false, name, ref flag, mutexSecurity); but there is another function mutex.SetAccessControl(mutexSecurity); which could be used instead.

What do you think? Should I create a PR for that?

**Comment by heebtob** [comment]
Okay I probably was to fast. 

mutex.SetAccessControl(mutexSecurity) is also .Net 4.8 and not available in .Net Standard 2.0.
I don't understand the Mutex Security in .Net Standard. Looks like there's no way to set it.
In .Net 5 there is a new Method MutexAcl.Create(Boolean, String, Boolean, MutexSecurity).

In the API Explorer I see the old Constructor Signature but in the Mutex itself the 4 arguments Constructor with MutexSecurity disappeared.

What a chaos 😄
Looks like there's no solution if the constructor is gone without a replacement.

**Comment by snakefoot** [comment]
Yes it looks like all applications must run with same service-account, or use .NetFramework. Or else the mutex-appender will not work across applications.

I guess you can try with `KeepFileOpen="false"` and `ConcurrentWrites="true"` but there will be a performance-hit (only 200 writes/sec). But if you include `<targets async="true">`  then the performance-hit will be reduced a lot (because of `AsyncTargetWrapper`)

But if having more than 2 application-instances writing to the same file with `KeepFileOpen="false"`, then logevents might get lost.

**Comment by snakefoot** [comment]
This compiles on Net5 when having `System.Threading.AccessControl` as dependency:

```c#
var mutexSecurity = new System.Security.AccessControl.MutexSecurity();
var everyoneSid = new System.Security.Principal.SecurityIdentifier(System.Security.Principal.WellKnownSidType.WorldSid, null);
mutexSecurity.AddAccessRule(new System.Security.AccessControl.MutexAccessRule(everyoneSid, System.Security.AccessControl.MutexRights.FullControl, System.Security.AccessControl.AccessControlType.Allow));
return System.Threading.MutexAcl.Create(false, "MutexName", out var _, mutexSecurity);
```

Guess my thoughts about splitting FileTarget into two, seems to be the only way for Net5.0. By having `FileTarget` and `ConcurrentFileTarget` See also #4244

**Comment by heebtob** [comment]
You were a minute faster than me 😄 

Actually I have a solution for my problem. I have a log setup function for every module. 
And because I only use .net 4.8 and .Net 5 I can always create the global mutex by myself with the access rights "everyone" before initializing NLog. I already tested this and it works.

```
        public static void checkSharableMutex(String filename)
        {
            if (System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
            {
                CreateSharableMutex("FileArchiveLock", filename);
                CreateSharableMutex("FileLock", filename);
            }
        }

        private static Mutex CreateSharableMutex(string mutexNamePrefix, string filename)
        {
            var name = GetMutexName(mutexNamePrefix, filename);

            return ForceCreateSharableMutex(name);
        }

        private static Mutex ForceCreateSharableMutex(string name)
        {
            if (System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
            {
                // Creates a mutex sharable by more than one process
                var mutexSecurity = new MutexSecurity();
                var everyoneSid = new SecurityIdentifier(WellKnownSidType.WorldSid, null);
                mutexSecurity.AddAccessRule(new MutexAccessRule(everyoneSid, MutexRights.FullControl, AccessControlType.Allow));

#if NET48
                // The constructor will either create new mutex or open
                // an existing one, in a thread-safe manner
                return new Mutex(false, name, out _, mutexSecurity);
#else
                return MutexAcl.Create(false, name, out _, mutexSecurity);
#endif
            }

            return new Mutex(false, name, out _);
        }

        private static string GetMutexName(string mutexNamePrefix, string filename)
        {
            const string mutexNameFormatString = @"Global\NLog-File{0}-{1}";
            const int maxMutexNameLength = 260;

            string canonicalName = Path.GetFullPath(filename).ToLowerInvariant();

            // Mutex names must not contain a slash, it's the namespace separator,
            // but all other are OK
            canonicalName = canonicalName.Replace('\\', '_');
            canonicalName = canonicalName.Replace('/', '_');
            string mutexName = string.Format(mutexNameFormatString, mutexNamePrefix, canonicalName);

            // A mutex name must not exceed MAX_PATH (260) characters
            if (mutexName.Length <= maxMutexNameLength)
            {
                return mutexName;
            }

            // The unusual case of the path being too long; let's hash the canonical name,
            // so it can be safely shortened and still remain unique
            string hash;
            using (MD5 md5 = MD5.Create())
            {
                byte[] bytes = md5.ComputeHash(Encoding.UTF8.GetBytes(canonicalName));
                hash = Convert.ToBase64String(bytes);
            }

            // The hash makes the name unique, but also add the end of the path,
            // so the end of the name tells us which file it is (for debugging)
            mutexName = string.Format(mutexNameFormatString, mutexNamePrefix, hash);
            int cutOffIndex = canonicalName.Length - (maxMutexNameLength - mutexName.Length);
            return mutexName + canonicalName.Substring(cutOffIndex);
        }
```

Many Thanks for your help!

**Comment by heebtob** [comment]
Yes it's quite a pain to have them all working correctly.
But .Net Framework and .Net 5 are definitely working together. what luck! 😄 

If there will be a ConcurrentFileTarget in the next version, I will definitely using that instead of my workaround.

Again many thanks for your help!

**Comment by snakefoot** [comment]
Also thank you for making a deep-dive into the wonderful world of compatibility issues between NetFramework and Net5.

Have revealed some places for improvement, though not sure if ConcurrentFileTarget-nuget-package will make it for NLog 5.0

**Comment by snakefoot** [comment]
Closing as resolved. Any further development should be reported to #4244

**Comment by snakefoot** [comment]
Resolved by #5770

**Comment by snakefoot** [comment]
NLog v6-preview1 has been released, that includes AtomicFileTarget, that supports ConcurrentWrites without using global-mutex:

- https://www.nuget.org/packages/NLog.Targets.AtomicFile

See also: https://nlog-project.org/2025/04/29/nlog-6-0-major-changes.html
